ethcore/sync/src/block_sync.rs

```diff
@@ -292,22 +312,31 @@ impl BlockDownloader {
 					let oldest_reorg = io.chain().pruning_info().earliest_state;
 					let last = self.last_imported_block;
 					if self.limit_reorg && best > last && (last == 0 || last < oldest_reorg) {
-						trace!(target: "sync", "No common block, disabling peer");
+						trace_sync!(self, "No common block, disabling peer");
 						return Err(BlockDownloaderImportError::Invalid);
 					}
 				}
 			},
 			State::Blocks => {
 				let count = headers.len();
-				// At least one of the heades must advance the subchain. Otherwise they are all useless.
+				// At least one of the headers must advance the subchain. Otherwise they are all useless.
 				if count == 0 || !any_known {
-					trace!(target: "sync", "No useful headers");
+					trace_sync!(self, "No useful headers, expected hash {:?}", expected_hash);
+					if let Some(eh) = expected_hash {
+						if self.blocks.heads_len() > 1 && !self.useless_expected_hashes.insert(eh) {
```

Comment:
I'm not entirely sure I understand this. This is to avoid importing the bad #57 block from which we won't be able to progress. But when we reset the sync it means we'll clear the `useless_expected_hashes`, what's preventing us from getting into the same bad fork again?

Reply:
The idea is that it should eventually find the correct blocks - it might try and reset a couple of times but once it gets a good peer it will move past the bad fork.

I clear `useless_expected_hashes` because I don't want to block further progress on a valid chain. It's possible to receive consecutive useless headers for some other reason other than being on the bad fork. The thinking is to be slightly overzealous on detecting being stuck (i.e. only 2 consecutive useless responses) but allowing it to keep trying again. Again on the assumption that it will retract and find the canon chain eventually. Happy to hear any better ideas though!

Reply:
The check is to prevent unnecessary resets when we're at the head of the chain (i.e. with just one 'subchain' head) with `NewBlocks` sync, which was happening often. I believe this is harmless anyway, but my approach here was to try and reset only when necessary.

The counter would probably work and is slightly simpler, but again might force some unnecessary resets.

Reply:
That makes sense. Can you leave a comment in the code here explaining that (the purpose of the `heads_len() > 1` check?). And perhaps it should check if the expected_hash is the last entry in `heads`, even if `heads_len()` is greater than 0.

If you think tracking useless responses by subchain head is preferable to a single counter, might I suggest changing `useless_expected_hashes` to a `H256FastMap<usize>` which tracks the number of useless responses per hash. That way, the reset threshold can be raised from 2 if necessary.

Regarding the counter,